core/src/main/scala/kafka/server/ControllerApis.scala

```diff
@@ -842,6 +848,114 @@ class ControllerApis(val requestChannel: RequestChannel,
       }
   }
 
+  def allowTokenRequests(request: RequestChannel.Request): Boolean = {
+    val protocol = request.context.securityProtocol
+    if (request.context.principal.tokenAuthenticated ||
+      // We allow forwarded requests to use PLAINTEXT for testing purposes
+      (request.isForwarded == false && protocol == SecurityProtocol.PLAINTEXT) ||
+      // disallow requests from 1-way SSL
+      (protocol == SecurityProtocol.SSL && request.context.principal == KafkaPrincipal.ANONYMOUS))
+      false
+    else
+      true
+  }
```

Comment:
Special-casing for PLAINTEXT, etc -- smells pretty bad.  I'm thinking the appropriate check here is just whether the request is forwarded or not.  Given that it takes cluster privileges to send an envelope request, we can trust the broker to perform these checks.  The controllers could conceivably be within some private network that only the brokers have access to, for example, in that case who knows what the authentication between broker and controller looks like.

Reply:
Yes, I agree that special casing PLAINTEXT is generally bad, and no production system should ever have PLAINTEXT. Here I am doing just what you suggested, I'm only allowing it on the forwarded cases and really that should only ever be on non production systems used for testing. The other checks still should not be allowed on the forwarded anyway.

Reply:
Updated with comment